src/core/stream_set.c

```diff
@@ -303,50 +368,59 @@ QuicStreamSetInitializeTransportParameters(
         UpdateAvailableStreams = TRUE;
     }
 
-    if (StreamSet->StreamTable != NULL) {
-        CXPLAT_HASHTABLE_ENUMERATOR Enumerator;
-        CXPLAT_HASHTABLE_ENTRY* Entry;
-        CxPlatHashtableEnumerateBegin(StreamSet->StreamTable, &Enumerator);
-        while ((Entry = CxPlatHashtableEnumerateNext(StreamSet->StreamTable, &Enumerator)) != NULL) {
-            QUIC_STREAM* Stream = CXPLAT_CONTAINING_RECORD(Entry, QUIC_STREAM, TableEntry);
+    CXPLAT_LIST_ENTRY* Link = StreamSet->WaitingStreams.Flink;
+    while (Link != &StreamSet->WaitingStreams) {
+        QUIC_STREAM* Stream =
+            CXPLAT_CONTAINING_RECORD(Link, QUIC_STREAM, WaitingLink);
+        Link = Link->Flink;
 
-            uint8_t FlowBlockedFlagsToRemove = 0;
+        const uint64_t StreamType = Stream->ID & STREAM_ID_MASK;
+        const uint64_t StreamIndex = (Stream->ID >> 2);
+        const QUIC_STREAM_TYPE_INFO* Info = &Stream->Connection->Streams.Types[StreamType];
 
-            uint64_t StreamType = Stream->ID & STREAM_ID_MASK;
-            uint64_t StreamCount = (Stream->ID >> 2) + 1;
-            const QUIC_STREAM_TYPE_INFO* Info =
-                &Stream->Connection->Streams.Types[StreamType];
-            if (Info->MaxTotalStreamCount >= StreamCount &&
-                Stream->OutFlowBlockedReasons & QUIC_FLOW_BLOCKED_STREAM_ID_FLOW_CONTROL) {
-                FlowBlockedFlagsToRemove |= QUIC_FLOW_BLOCKED_STREAM_ID_FLOW_CONTROL;
-                QuicStreamIndicatePeerAccepted(Stream);
-            } else {
-                QuicSendSetSendFlag(
-                    &Stream->Connection->Send,
-                    STREAM_ID_IS_UNI_DIR(Stream->ID) ?
-                        QUIC_CONN_SEND_FLAG_UNI_STREAMS_BLOCKED : QUIC_CONN_SEND_FLAG_BIDI_STREAMS_BLOCKED);
-            }
+        CXPLAT_DBG_ASSERT(Stream->OutFlowBlockedReasons & QUIC_FLOW_BLOCKED_STREAM_ID_FLOW_CONTROL);
 
-            uint64_t NewMaxAllowedSendOffset =
-                QuicStreamGetInitialMaxDataFromTP(
-                    Stream->ID,
-                    QuicConnIsServer(Connection),
-                    &Connection->PeerTransportParams);
+        uint8_t FlowBlockedFlagsToRemove = 0;
+        if (StreamIndex < Info->MaxTotalStreamCount) {
```

Comment:
`WaitingStreams` is an ordered list, right? Assuming that, if this condition fails, we should simply `break` out of the loop instead of continuing to walk the list.
```suggestion
        if (StreamIndex >= Info->MaxTotalStreamCount) {
            break;
        }
```

Reply:
It is ordered but all types of streams are in the same list so we need to iterate at least up to the maximum "MaxTotalStreamCount" for unidirectional and bidirectional streams.

The values are already available in the function, I'll add that early exit check.

Reply:
Hum, that change broke the test. Looking closer.

Reply:
Ugh, now I remember why there was no early exit on that loop.
For every stream that is can't be opened with the peer yet, we still need to indicate that the stream is blocked + all that is below in the loop.

For this loop, we do need to iterate through all waiting streams.